### ISSUE TITLE
Serialization error in testing

### ISSUE THREAD

**Issue by giovannipizzi** [aiidateam/aiida_core#2363]
In one of the final phases of the tests on travis, this random error occurred:
```
Submitting a WorkChain which contains a workfunction.
12/21/2018 08:38:12 PM <31656> aiida.orm.node.process.workflow.workchain.WorkChainNode: [ERROR] Exception calling cleanup method <functools.partial object at 0x7f05eadf2260>
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/plumpy/processes.py", line 758, in on_close
    cleanup()
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/kiwipy/rmq/communicator.py", line 582, in remove_rpc_subscriber
    return self._run_task(coro)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/kiwipy/rmq/communicator.py", line 677, in _run_task
    return self.tornado_to_kiwi_future(self._create_task(coro)).result(timeout=self.TASK_TIMEOUT)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/kiwipy/rmq/communicator.py", line 656, in tornado_to_kiwi_future
    self.loop().add_future(tornado_future, done)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/tornado/ioloop.py", line 597, in add_future
    lambda future: self.add_callback(callback, future))
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/tornado/concurrent.py", line 270, in add_done_callback
    self._callbacks.append(fn)
AttributeError: 'NoneType' object has no attribute 'append'
Traceback (most recent call last):
  File "/home/travis/build/aiidateam/aiida_core/aiida/common/log.py", line 75, in emit
    orm.Log.objects(backend).create_entry_from_record(record)
  File "/home/travis/build/aiidateam/aiida_core/aiida/orm/logs.py", line 70, in create_entry_from_record
    metadata=metadata)
  File "/home/travis/build/aiidateam/aiida_core/aiida/orm/logs.py", line 109, in __init__
    metadata=metadata)
  File "/home/travis/build/aiidateam/aiida_core/aiida/orm/implementation/backends.py", line 179, in create
    return self.ENTITY_CLASS(backend=self._backend, **kwargs)  # pylint: disable=not-callable
  File "/home/travis/build/aiidateam/aiida_core/aiida/orm/implementation/django/logs.py", line 36, in __init__
    metadata=json.dumps(metadata))
  File "/home/travis/build/aiidateam/aiida_core/aiida/common/json.py", line 47, in dumps
    return simplejson.dumps(data, ensure_ascii=False, encoding='utf8', **kwargs)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/simplejson/__init__.py", line 399, in dumps
    **kw).encode(obj)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/simplejson/encoder.py", line 296, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/simplejson/encoder.py", line 378, in iterencode
    return _iterencode(o, 0)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/simplejson/encoder.py", line 273, in default
    o.__class__.__name__)
TypeError: Object of type partial is not JSON serializable
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.14/bin/verdi", line 11, in <module>
    sys.exit(verdi())
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/home/travis/build/aiidateam/aiida_core/aiida/cmdline/utils/decorators.py", line 72, in decorated_function
    return function(*args, **kwargs)
  File "/home/travis/build/aiidateam/aiida_core/aiida/cmdline/commands/cmd_run.py", line 112, in run
    exec(compile(handle.read(), scriptname, 'exec'), globals_dict)  # yapf: disable # pylint: disable=exec-used
  File "/home/travis/build/aiidateam/aiida_core/.ci/test_daemon.py", line 463, in <module>
    main()
  File "/home/travis/build/aiidateam/aiida_core/.ci/test_daemon.py", line 362, in main
    pk = submit(WorkFunctionRunnerWorkChain, input=value).pk
  File "/home/travis/build/aiidateam/aiida_core/aiida/work/launch.py", line 89, in submit
    process.close()
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/plumpy/processes.py", line 863, in close
    call_with_super_check(self.on_close)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/plumpy/base/utils.py", line 29, in call_with_super_check
    fn(*args, **kwargs)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/plumpy/base/utils.py", line 16, in new_fn
    fn(self, *args, **kwargs)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/plumpy/processes.py", line 760, in on_close
    self.logger.exception('Exception calling cleanup method %s', cleanup)
  File "/opt/python/2.7.14/lib/python2.7/logging/__init__.py", line 1473, in exception
    self.logger.error(msg, *args, **kwargs)
  File "/opt/python/2.7.14/lib/python2.7/logging/__init__.py", line 1193, in error
    self._log(ERROR, msg, args, **kwargs)
  File "/opt/python/2.7.14/lib/python2.7/logging/__init__.py", line 1286, in _log
    self.handle(record)
  File "/opt/python/2.7.14/lib/python2.7/logging/__init__.py", line 1296, in handle
    self.callHandlers(record)
  File "/opt/python/2.7.14/lib/python2.7/logging/__init__.py", line 1336, in callHandlers
    hdlr.handle(record)
  File "/opt/python/2.7.14/lib/python2.7/logging/__init__.py", line 759, in handle
    self.emit(record)
  File "/home/travis/build/aiidateam/aiida_core/aiida/common/log.py", line 75, in emit
    orm.Log.objects(backend).create_entry_from_record(record)
  File "/home/travis/build/aiidateam/aiida_core/aiida/orm/logs.py", line 70, in create_entry_from_record
    metadata=metadata)
  File "/home/travis/build/aiidateam/aiida_core/aiida/orm/logs.py", line 109, in __init__
    metadata=metadata)
  File "/home/travis/build/aiidateam/aiida_core/aiida/orm/implementation/backends.py", line 179, in create
    return self.ENTITY_CLASS(backend=self._backend, **kwargs)  # pylint: disable=not-callable
  File "/home/travis/build/aiidateam/aiida_core/aiida/orm/implementation/django/logs.py", line 36, in __init__
    metadata=json.dumps(metadata))
  File "/home/travis/build/aiidateam/aiida_core/aiida/common/json.py", line 47, in dumps
    return simplejson.dumps(data, ensure_ascii=False, encoding='utf8', **kwargs)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/simplejson/__init__.py", line 399, in dumps
    **kw).encode(obj)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/simplejson/encoder.py", line 296, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/simplejson/encoder.py", line 378, in iterencode
    return _iterencode(o, 0)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/simplejson/encoder.py", line 273, in default
    o.__class__.__name__)
TypeError: Object of type partial is not JSON serializable
```

Full log: 
[log.txt](https://github.com/aiidateam/aiida_core/files/2714034/log.txt)

**Comment by sphuber** [comment]
The problem here probably is that the `metadata` of the log record contains a value that is a partial object, i.e. a function wrapped by `functools.partial`. I have found one example where an exception occurs in a cleanup method registered for a process, which are called in `on_close` in `plumpy.Process`. This method will call `self.logger.exception('Exception calling cleanup method %s:', cleanup)`. If the `cleanup` is a partial, the `metadata` of the log record that is created will contain in the `args` key a list of arguments, one of which will the partial object. I think checking for these in the `orm.Log.create_from_record` and dereferencing the wrapped function should work.

**Comment by sphuber** [comment]
I found another occasion that will trigger this problem, where it is not a `partial` but a bound method that will be in the args and therefore fail the serialization. I am not sure if we can find a robust way too guarantee that the args will always be serializable so maybe we should just drop them? We are already dropping the `exc_info` from the `record` for that very reason.

**Comment by sphuber** [comment]
Tentatively fixed in PR #2469
